raft_test.go

```diff
@@ -1019,6 +1020,79 @@ func TestRaft_SnapshotRestore(t *testing.T) {
 	}
 }
 
+func TestRaft_SnapshotRestore_Monotonic(t *testing.T) {
+	// Make the cluster
+	conf := inmemConfig(t)
+	conf.TrailingLogs = 10
+	opts := &MakeClusterOpts{
+		Peers:         1,
+		Bootstrap:     true,
+		Conf:          conf,
+		MonotonicLogs: true,
+	}
+	c := MakeClusterCustom(t, opts)
+	defer c.Close()
+
+	leader := c.Leader()
+
+	// Commit a lot of things
+	var future Future
+	for i := 0; i < 100; i++ {
+		future = leader.Apply([]byte(fmt.Sprintf("test%d", i)), 0)
+	}
+
+	// Wait for the last future to apply
+	if err := future.Error(); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// Take a snapshot
+	snapFuture := leader.Snapshot()
+	if err := snapFuture.Error(); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// Check for snapshot
+	snaps, _ := leader.snapshots.List()
+	if len(snaps) != 1 {
+		t.Fatalf("should have a snapshot")
+	}
+	snap := snaps[0]
+
+	// Logs should be trimmed
+	if idx, _ := leader.logs.FirstIndex(); idx != snap.Index-conf.TrailingLogs+1 {
+		t.Fatalf("should trim logs to %d: but is %d", snap.Index-conf.TrailingLogs+1, idx)
+	}
+
+	// Shutdown
+	shutdown := leader.Shutdown()
+	if err := shutdown.Error(); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// Restart the Raft
+	r := leader
+	// Can't just reuse the old transport as it will be closed
+	_, trans2 := NewInmemTransport(r.trans.LocalAddr())
+	cfg := r.config()
+	r, err := NewRaft(&cfg, r.fsm, r.logs, r.stable, r.snapshots, trans2)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	c.rafts[0] = r
+
+	// We should have restored from the snapshot!
+	if last := r.getLastApplied(); last != snap.Index {
+		t.Fatalf("bad last index: %d, expecting %d", last, snap.Index)
+	}
+
+	// Verify that logs have been reset
+	first, _ := r.logs.FirstIndex()
+	last, _ := r.logs.LastIndex()
+	assert.Zero(t, first)
+	assert.Zero(t, last)
```

Comment:
Does this imply that we are reset the WAL Index back to 0 after a successful snapshot restore? What will be the index of the next value i write through raft?

Reply:
I believe the next index would just be written as usual. A log entry can be appended to any index in an empty log. We just have to clear the stale entries from the log (post-restore) so that we don't violate raft-wal's assertion of a gapless log.

Reply:
@briankassouf great question but no it doesn't imply that. As Mike said, this is just validating that the LogStore now contains zero logs (and therefore first and last are no longer meaningful concepts).

The first write to an empty LogStore doesn't necessarily have to be index 1 since a follower that starts after the leader has already taken at least one snapshot will never experience that.

In the case of snapshots, the current behaviour of this library is to artificially set the "index" of the snapshot it's restoring to the current `LastIndex + 1` which ensures it's larger than any index every actually written to this cluster (and keeps the term). So the first Log written after the restore (the NoOp log used to force snapshot replication to followres) will always be at `LastIndex + 1` of the current log, and that will be the new first log stored in each node's LogStore with this change.

Reply:
Good catch Brian. This actually indicates one place where we should _not_ change the behaviour. Will add comments above.

Reply:
So this test should _not_ touch the logs at all - we are just restoring from local disk not replacing the local state with external snapshot state.

We would however need a test that exercises the path that the Vault integration tests do:
 - setup  a cluster with at least one follower (probably 3 nodes)
 - write some stuff
 - take a snapshot and store it somewhere
 - write some more stuff
 - restore the snapshot on leader
 - validate that the leader and followers have only the restored data and not the extra stuff written in between...

That is the test that exercises the issues that this PR is designed to fix I think?

Reply:
Makes sense. Should be easy enough to add. I will have a bit of time to work on this later today. I'll ping you for re-review then. Sorry for the churn on this. Hopefully the next pass is good to go!